project/frontend/src/components/ImageSlider.js

```diff
@@ -0,0 +1,186 @@
+import './ImageSlider.css';
+
+import React, { useState, useReducer } from 'react';
+import { useSwipeable } from 'react-swipeable';
+
+import Next from '../assets/next.svg';
+import Prev from '../assets/prev.svg';
+
+/**
+ * Types of swipe actions.
+ * @enum {number}
+ */
+const Action = {
+  SWIPING: 0,
+  SWIPED_LEFT: 1,
+  SWIPED_RIGHT: 2,
+  PREV_SLIDE: 3,
+  NEXT_SLIDE: 4,
+};
+
+/**
+ * The default width of a slide.
+ *
+ * @type {number}
+ */
+const width = 344;
+
+/**
+ * A reducer for the ImageSlider component. Updates ImageSlider state
+ * based on the previous state and a specified action.
+ *
+ * @param {!Object<string, *>} previous The previous state.
+ * @param {!Action} action An action type that may determine how the
+ *     state is updated.
+ */
+function reducer(previous, action) {
+  let { left, lastLeft, slide, slideCount, shifting } = previous;
+
+  switch (action.type) {
+    case Action.SWIPING:
+      const minLeft = (slideCount - 1) * -width;
+      const maxLeft = 0.0;
+      left = Math.min(Math.max(lastLeft + action.deltaX, minLeft), maxLeft);
+      shifting = false;
+
+      // Update the current slide if user slides
+      // passed multiple slides in one motion.
+      const dir = action.deltaX > 0 ? 'right' : 'left';
+      if (dir === 'left') {
+        if (left <= (slide + 1) * -width) {
+          slide++;
+        } else if (left > slide * -width) {
+          slide--;
+        }
+      } else {
+        if (left >= (slide - 1) * -width) {
+          slide--;
+        } else if (left < slide * -width) {
+          slide++;
+        }
+      }
+      break;
+    case Action.SWIPED_LEFT:
+      // Spring slider back to current slide or to next
+      // slide depending on how far the user slid.
+      if (slide < slideCount - 1) {
+        if (Math.abs(left) % width > width / 2 || action.velocity >= 0.5) {
+          slide++;
+        }
+      }
+      break;
+    case Action.SWIPED_RIGHT:
+      // Spring slider back to current slide or to previous
+      // slide depending on how far the user slid.
+      if (slide > 0) {
+        if (
+          width - (Math.abs(left) % width) > width / 2 ||
+          action.velocity >= 0.5
+        ) {
+          slide--;
+        }
+      }
+      break;
+    case Action.PREV_SLIDE:
+      if (slide > 0) slide--;
+      break;
+    case Action.NEXT_SLIDE:
+      if (slide < slideCount - 1) slide++;
+      break;
+    default:
+      throw new Error();
+  }
+
+  if (action.type !== Action.SWIPING) {
+    left = slide * -width;
+    lastLeft = left;
+    shifting = true;
+  }
+
+  return { left, lastLeft, slide, slideCount, shifting };
+}
```

Comment:
nit: (for a follow up CL) move the reducer logic out into it's own file ImageSliderReducer.js and put all your reducers in their own directory

Reply:
I'll add a todo.